### PR TITLE
pkg: refine tls && update tidb version

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 

1. need to support online reload for tls https://github.com/pingcap/dm/issues/467
2. need to support CN certification https://github.com/pingcap/dm/issues/524

### What is changed and how it works?
1. refine pkg for tls, for example, provide some function for grpc to use reload tls file, most code is copied from  https://github.com/pingcap/tidb-lightning/blob/master/lightning/common/security.go
2. add certification for Common Name in CA

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: pkg/utils/security.go [hunk 1]
```diff
@@ -17,41 +17,170 @@ import (
...
-			// Load the client certificates from disk
-			certificate, err := tls.LoadX509KeyPair(SSLCert, SSLKey)
-			if err != nil {
-				return nil, errors.Errorf("could not load client key pair: %s", err)
-			}
-			certificates = append(certificates, certificate)
-		}
+func ToTLSConfig(caPath, certPath, keyPath string) (*tls.Config, error) {
+	return ToTLSConfig(caPath, certPath, keyPath)
+}
```
Comment (kennytm) [inline pkg/utils/security.go]: ????????

Comment (WangXiangUSTC) [reply pkg/utils/security.go]: a mistake, should be `ToTLSConfigWithVerify(caPath, certPath, keyPath, nil)`

**Comment by WangXiangUSTC** [comment]
@kennytm @csuzhangxc PTAL

File: pkg/utils/security.go [hunk 2]
```diff
@@ -17,41 +17,170 @@ import (
...
+		ClientCAs:    certPool,
+		NextProtos:   []string{"h2", "http/1.1"}, // specify `h2` to let Go use HTTP/2.
+	}
+
+	checkCN := make(map[string]struct{})
+	for _, cn := range verifyCN {
+		cn = strings.TrimSpace(cn)
+		checkCN[cn] = struct{}{}
+	}
+	if len(checkCN) != 0 {
```
Comment (kennytm) [inline pkg/utils/security.go]: move lines 80-84 inside the `if` branch.

```suggestion
	if len(verifyCN) != 0 {
		checkCN := make(map[string]struct{})
		for _, cn := range verifyCN {
			cn = strings.TrimSpace(cn)
			checkCN[cn] = struct{}{}
		}
```

File: pkg/utils/security.go [hunk 3]
```diff
@@ -17,41 +17,170 @@ import (
...
+		tlsCfg.ClientAuth = tls.RequireAndVerifyClientCert
+
+		tlsCfg.VerifyPeerCertificate = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			if len(checkCN) == 0 {
+				return nil
+			}
+
```
Comment (kennytm) [inline pkg/utils/security.go]:
```suggestion
```

File: pkg/utils/security.go [hunk 4]
```diff
@@ -17,41 +17,170 @@ import (
...
+	client := ClientWithTLS(inner)
+
+	return &TLS{
+		inner:  inner,
+		client: client,
+		url:    "https://" + host,
+	}, nil
+}
+
+// ClientWithTLS ...
```
Comment (kennytm) [inline pkg/utils/security.go]: fill in the `...` 🙂

File: pkg/utils/tls_test/ca.csr [hunk 5]
```diff
@@ -0,0 +1,7 @@
+-----BEGIN CERTIFICATE REQUEST-----
```
Comment (kennytm) [inline pkg/utils/tls_test/ca.csr]: no need to include the CSR files.

but perhaps add the instruction how to regenerate the *.key and *.pem

File: pkg/utils/security_test.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	resp, err := ClientWithTLS(clientTLS1).Get(url)
+	c.Assert(err, IsNil)
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	c.Assert(err, IsNil)
+	c.Assert(string(body), Equals, "This an example server")
+
+	// client2 can't visit server
+	_, err = ClientWithTLS(clientTLS2).Get(url)
+	c.Assert(err, ErrorMatches, ".*tls: bad certificate")
```
Comment (kennytm) [inline pkg/utils/security_test.go:138]: will the server accept `client2` if the common name is `localhost` (the CA's common name)?

File: pkg/utils/security.go (hunk 2 above)
Comment (WangXiangUSTC) [reply pkg/utils/security.go]: refine in dea0b3a

File: pkg/utils/security.go (hunk 3 above)
Comment (WangXiangUSTC) [reply pkg/utils/security.go]: removed in dea0b3a

File: pkg/utils/security.go (hunk 4 above)
Comment (WangXiangUSTC) [reply pkg/utils/security.go]: fill in dea0b3a

File: pkg/utils/tls_test/ca.csr (hunk 5 above)
Comment (WangXiangUSTC) [reply pkg/utils/tls_test/ca.csr]: do you mean:
1. delete the CSR files
2. add some comments in *.key and *.pem to introduce how to generate these files

File: pkg/utils/security_test.go (hunk 6 above)
Comment (WangXiangUSTC) [reply pkg/utils/security_test.go:138]: no, you will see the log of unit test:
```
2020/03/12 12:21:44 http: TLS handshake error from 127.0.0.1:51864: client certificate authentication failed. The Common Name from the client certificate [client2 localhost] was not found in the configuration cluster-verify-cn with value: [client1]
```

the common name already include `localhost`

File: pkg/utils/tls_test/ca.csr (hunk 5 above)
Comment (kennytm) [reply pkg/utils/tls_test/ca.csr]: 1. delete irrelevant files like the CSR and SRL files.
2. add a readme or a *.sh file about how we can regenerate the pem & key for the test.

Comment (WangXiangUSTC) [reply pkg/utils/tls_test/ca.csr]: ok, I do it in b91ec85

**Review by kennytm — COMMENTED** [review]
LGTM

**Review by july2993 — APPROVED** [review]

**Pull request merged**